### DISCUSSION TITLE
Is it possible to position a png image inside a frame which is already positioned in the window using ".center_of"

### DISCUSSION THREAD

**Discussion by santokalayil** [fltk-rs/fltk-rs#676] [Q&A]
Is it possible to position a png image inside a frame which is already positioned in the window using ".center_of" method? that width:height proportion of the png image should not change as well. If I manually set the x position, the problem is that when resizing the window, it is no more in the center position.

See here. I cannot position my logo in window..
![image](https://user-images.githubusercontent.com/49450970/118115385-7de69900-b406-11eb-8968-d1e7437abf06.png)

I need to place like this in the following image:
![image](https://user-images.githubusercontent.com/49450970/118115919-31e82400-b407-11eb-8f5b-efd707df6618.png)

"H" needs to replaced by that png image.

**Comment by MoAlyousef** [comment]
You can try drawing the image instead of using set_image().
```rust
let mut frame = /* construct a frame */;
frame.draw(|f| {
    let png = /* get image */;
    png.scale(f.w(), f.h(), true, true); // should rescale the image with redraws of the frame (for ex. with resizing)
    png.draw(f.x(), f.y(), f.w(), f.h());
});
```

**Reply by santokalayil** [reply]
I used the same code in the following way. still the same issue.
```
impl LogoImage {
    pub fn new(x:i32, y:i32, w:i32, h:i32, label:&'static str) -> LogoImage {
        let mut group = Group::new(x, y, w, h, label);
        group.set_align(fltk::enums::Align::Center);
        group.set_frame(fltk::enums::FrameType::FlatBox);
        group.set_color(fltk::enums::Color::Red);

        let mut logo_frame = Frame::default()
        .with_size(50, 50)
        .center_of(&group)
        .with_label("H");
        logo_frame.set_frame(fltk::enums::FrameType::FlatBox);
        logo_frame.set_color(Color::Yellow);

        logo_frame.draw(|f| {
            // let image_data = read("assets\\app_icon.png").unwrap();
            let mut img = PngImage::load("assets\\app_icon.png").unwrap();
            img.scale(f.w(), f.h(), true, true);
            img.draw(f.x(), f.y(), f.w(), f.h());
        });
        
        group.end();
        LogoImage { group, logo_frame}
    }
}
```

![image](https://user-images.githubusercontent.com/49450970/118164318-9673a500-b440-11eb-8175-d79213b33286.png)

but got this only.. earlier also I have seen your code and done in the same way.. but it did not work..

**Reply by MoAlyousef** [reply]
It depends on where you need to place it exactly within the frame. You can modify the img.draw() coordinates, also you can select to scale the image where you don’t preserve the aspect ratio.

**Reply by santokalayil** [reply]
I tried everything.. only statically tweaking the x coordinates worked but while resizing it loses it's center position. Also I have given same width and height (200:200) to the LogoImage group but it scales to match the width of other elements in the parent widget. Therefore in effect, the width and height are a lot different (the width gets more than double size of the height. Also it changes hugely different if I apply 50:50).
The Repository Link: [Ecclessiastica App](https://github.com/santokalayil/Ecclessiastica_App_FLTK_Rust)
Please find the code in src/views/logo.rs file. LogoImage is called in login.rs file in the same views directory.

**Reply by eadf** [reply]
Try this
```
logo_frame.draw(|f| {
    let mut img = PngImage::load( std::path::Path::new("assets/app_icon.png")).unwrap();
    img.scale(f.w(), f.h(), true, true);
    img.draw(f.x() + f.w()/2 - img.width()/2,
                     f.y() + f.h()/2 - img.height()/2,
                     img.width(), 
                     img.height());
 });
```

Edit: removed superfluous `.borrow()`.

**Reply by santokalayil** [reply]
It perfectly worked... Thank you very much mate...

**Comment by santokalayil** [comment]
logo_frame.draw(|f| {
    let mut img = PngImage::load( std::path::Path::new("assets/app_icon.png")).unwrap();
    img.scale(f.w(), f.h(), true, true);
    img.draw(f.x() + f.w()/2 - img.width()/2,
                     f.y() + f.h()/2 - img.height()/2,
                     img.width(), 
                     img.height());
 });

**Comment by eadf** [comment]
Just looked at this again. You probably don't want to load the image from disk every time you `draw()` it: 
Quick fix:

```rust
 logo_frame.draw({
      let mut img = PngImage::load(std::path::Path::new("assets/app_icon.png")).unwrap();
      move |f| {
          img.scale(f.w(), f.h(), true, true);
          img.draw(
              f.x() + f.w() / 2 - img.width() / 2,
              f.y() + f.h() / 2 - img.height() / 2,
              img.width(),
              img.height(),
          );
      }
  });
```
`img.scale(f.w(), f.h(), true, true);` could also be moved outside of the `|f|{}` if the frame size is static.

**Reply by wyhinton** [reply]
so adding the ```move|f|``` closure inside the draw function means that any code outside of that closure is only called once, and the code inside that ```move|f|``` closure is called on every redraw?

**Reply by eadf** [reply]
Yes, the closure `|f|` is just an anonymous function, anything inside it will be executed for every draw(). (unless the compiler is smart enough to figure out some optimizations).
The `let mut img=..;` part could just as well be placed outside `logo_frame.draw(`. I just think it's neat to group the 'moved' data together with the closure that uses it.
